pandas/core/groupby/grouper.py

```diff
@@ -568,7 +568,9 @@ def codes(self) -> np.ndarray:
     @cache_readonly
     def result_index(self) -> Index:
         if self.all_grouper is not None:
-            return recode_from_groupby(self.all_grouper, self.sort, self.group_index)
+            group_idx = self.group_index
+            assert isinstance(group_idx, CategoricalIndex)  # set in __init__
+            return recode_from_groupby(self.all_grouper, self.sort, group_idx)
```

Comment:
is this needed.

are the type annotations for `def group_index(self) -> Index` and `_group_index: Optional[Index] = None` correct.

in \_\_init__

```
self._group_index = CategoricalIndex(..
```
or 
```
(
                self.grouper,
                self._codes,
                self._group_index,
            ) = index._get_grouper_for_level(self.grouper, level)

```
so self._group_index  can only be CategoricalIndex or None and group_index can only be CategoricalIndex ?

Reply:
IIRC its the self.all_grouper check a few lines up that ensures we have a CategoricalIndex here

Reply:
self.all_grouper is not None when is_categorical_dtype(self.grouper) so I don't think that narrows it. (but I may need to trace through further)